### ISSUE TITLE
[REVIEW]: Interface to high-performance periodic coupled-cluster theory calculations with atom-centered, localized basis functions

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4040]
**Submitting author:** @Evmoerman (<a href="http://orcid.org/0000-0002-9725-612X">Evgeny Moerman</a>)
**Repository:** https://gitlab.com/moerman1/fhi-cc4s
**Branch with paper.md** (empty if default branch): 
**Version:** 1.0.1
**Editor:** @rkurchin
**Reviewers:** @awvwgk, @wcwitt
**Archive:** 10.5281/zenodo.6658117

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/ce4d17e7211b3eb8b4f7472c18b1482a/status.svg)](https://joss.theoj.org/papers/ce4d17e7211b3eb8b4f7472c18b1482a)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/ce4d17e7211b3eb8b4f7472c18b1482a"><img src="https://joss.theoj.org/papers/ce4d17e7211b3eb8b4f7472c18b1482a/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/ce4d17e7211b3eb8b4f7472c18b1482a/status.svg)](https://joss.theoj.org/papers/ce4d17e7211b3eb8b4f7472c18b1482a)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://gitlab.com/moerman1/fhi-cc4s" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@awvwgk & @wcwitt, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @rkurchin know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @awvwgk
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://gitlab.com/moerman1/fhi-cc4s">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/moerman1/fhi-cc4s">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@Evmoerman) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
  - Fails to compile: https://gitlab.com/moerman1/fhi-cc4s/-/issues/1
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
   - https://gitlab.com/moerman1/fhi-cc4s/-/issues/4
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
   - https://gitlab.com/moerman1/fhi-cc4s/-/issues/5 
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support
   - https://gitlab.com/moerman1/fhi-cc4s/-/issues/3

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @wcwitt
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://gitlab.com/moerman1/fhi-cc4s">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/moerman1/fhi-cc4s">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@Evmoerman) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @awvwgk, @wcwitt it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1061

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.27 s (230.0 files/s, 51273.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Fortran 90                      30           1725           1748           6339
TeX                              5            147              0           1406
Markdown                        19            219              0           1386
Bourne Shell                     1             54             92            169
YAML                             3             26              5            106
make                             3             64             18             97
-------------------------------------------------------------------------------
SUM:                            61           2235           1863           9503
-------------------------------------------------------------------------------


Statistical information for the repository 'd500ae595a303c04f7b87f2a' was
gathered on 2022/01/07.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- 10.3389/fmats.2019.00123 may be a valid DOI for title: Coupled cluster theory in materials science
- 10.1016/j.cpc.2009.06.022 may be a valid DOI for title: Ab initio molecular simulations with numeric atom-centered orbitals
- 10.1103/physrev.46.618 may be a valid DOI for title: Note on an approximation treatment for many-electron systems
- 10.1063/1.3043729 may be a valid DOI for title: The ground state correlation energy of the random phase approximation from a ring coupled cluster doubles approach
- 10.1016/s0009-2614(89)87395-6 may be a valid DOI for title: A fifth-order perturbation comparison of electron correlation theories
- 10.1063/1.455269 may be a valid DOI for title: An efficient reformulation of the closed-shell coupled cluster single and double excitation (CCSD) equations
- 10.1063/1.4977994 may be a valid DOI for title: Low rank factorization of the Coulomb integrals for periodic coupled cluster theory
- 10.1088/1367-2630/14/5/053020 may be a valid DOI for title: Resolution-of-identity approach to Hartree–Fock, hybrid density functionals, RPA, MP2 and GW with numeric atom-centered orbital basis functions
- 10.1021/acs.jpclett.0c00481.s001 may be a valid DOI for title: Accuracy of Localized Resolution of the Identity in Periodic Hybrid Functional Calculations with Numerical Atomic Orbitals
- 10.1021/acs.jpclett.8b03679.s001 may be a valid DOI for title: Physisorption of water on graphene: subchemical accuracy from many-body electronic structure methods
- 10.1063/1.5055706 may be a valid DOI for title: Reaction energetics of hydrogen on Si (100) surface: A periodic many-electron theory study
- 10.1103/physrevb.101.165138 may be a valid DOI for title: Electronic structure of bulk manganese oxide and nickel oxide from coupled cluster theory
- 10.1021/acs.jctc.7b00049 may be a valid DOI for title: Gaussian-based coupled-cluster theory for the ground-state and band structure of solids
- 10.1021/acs.jctc.5b00422 may be a valid DOI for title: Can single-reference coupled cluster theory describe static correlation?
- 10.1103/physrevlett.117.133002 may be a valid DOI for title: Towards efficient orbital-dependent density functionals for weak and strong correlation
- 10.1021/acs.jpca.9b03976.s001 may be a valid DOI for title: The CUAGAU Set of Coupled-Cluster Reference Data for Small Copper, Silver, and Gold Compounds and Assessment of DFT Methods
- 10.1063/1.1388045 may be a valid DOI for title: Extended benchmark studies of coupled cluster theory through triple excitations
- 10.1063/1.4985878 may be a valid DOI for title: Properties of the water to boron nitride interaction: From zero to two dimensions with benchmark accuracy
- 10.1103/physrevb.101.241113 may be a valid DOI for title: First-principles coupled cluster theory of the electronic spectrum of transition metal dichalcogenides
- 10.1103/physrevlett.52.1830 may be a valid DOI for title: Transition-metal monoxides: band or Mott insulators
- 10.1103/physrevb.54.1703 may be a valid DOI for title: Separable dual-space Gaussian pseudopotentials
- 10.1103/physrevb.73.035404 may be a valid DOI for title: Comparison of the full-potential and frozen-core approximation approaches to density-functional calculations of surfaces
- 10.1016/0022-3093(95)00355-x may be a valid DOI for title: Ab initio molecular dynamics for liquid metals
- 10.1039/9781849734790-00168 may be a valid DOI for title: On choosing the best density functional approximation
- 10.1007/128_2014_600 may be a valid DOI for title: Judging density-functional approximations: Some pitfalls of statistics
- 10.1021/acs.jctc.9b00957.s001 may be a valid DOI for title: Integral-direct and parallel implementation of the CCSD (T) method: Algorithmic developments and large-scale applications
- 10.1021/acs.jctc.9b00511.s001 may be a valid DOI for title: Approaching the basis set limit of CCSD (T) energies for large molecules with local natural orbital coupled-cluster methods
- 10.1021/acs.jctc.0c01129.s001 may be a valid DOI for title: Scalable Electron Correlation Methods. 8. Explicitly Correlated Open-Shell Coupled-Cluster with Pair Natural Orbitals PNO-RCCSD (T)-F12 and PNO-UCCSD (T)-F12
- 10.1063/1.464746 may be a valid DOI for title: The equation of motion coupled-cluster method. A systematic biorthogonal approach to molecular excitation energies, transition probabilities, and excited state properties
- 10.1103/physrevb.59.1758 may be a valid DOI for title: From ultrasoft pseudopotentials to the projector augmented-wave method
- 10.1088/1367-2630/17/9/093020 may be a valid DOI for title: Accurate localized resolution of identity approach for linear-scaling hybrid density functionals and for many-body perturbation theory
- 10.1002/jcc.23284 may be a valid DOI for title: Attractive electron–electron interactions within robust local fitting approximations
- 10.1063/1.1679012 may be a valid DOI for title: Coulombic potential energy integrals and approximations
- 10.1016/0009-2614(93)87156-w may be a valid DOI for title: Use of approximate integrals in ab initio theory. An application in MP2 energy calculations
- 10.1021/acs.jctc.0c00101 may be a valid DOI for title: Excitons in solids from periodic equation-of-motion coupled-cluster theory

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04040/joss.04040/10.21105.joss.04040.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04040/joss.04040/10.21105.joss.04040.pdf) :page_facing_up: :point_left:

**Comment by rkurchin** [comment]
@Evmoerman please do add in those missing DOI's when you get a chance!

**Comment by awvwgk** [comment]
I guess the functionality review will heavily depend whether we or our home institutes have bought a license for [FHI-aims](https://fhi-aims.org/), also the other end [CC4S](https://cqc.itp.tuwien.ac.at/code.html) seems a bit difficult to get hands on.

**Comment by Evmoerman** [comment]
@rkurchin Sure, I will take care of the DOIs today or tomorrow.
@awvwgk I have direct contact to the maintainers/owners of both codes. Hence, if you tell me exactly to which institution/individual access to CC4S/FHI-aims should be given, this can be easily arranged.

**Comment by awvwgk** [comment]
@Evmoerman Thanks for the offer, I was just about to write Volker Blum whether I can get access to FHI-aims in some way. I'm currently in Bonn at the Mulliken Center for Theoretical Chemistry, in [Stefan Grimme's group](https://www.chemie.uni-bonn.de/pctc/mulliken-center/grimme/).

**Comment by Evmoerman** [comment]
@rkurchin Missing DOIs have been added.
@awvwgk Emails have been sent to the respective people in charge. Hopefully, by next week you will have access to both codes.

**Comment by Evmoerman** [comment]
@whedon generate pdf

**Comment by Evmoerman** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04040/joss.04040/10.21105.joss.04040.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04040/joss.04040/10.21105.joss.04040.pdf) :page_facing_up: :point_left:

**Comment by Evmoerman** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.3389/fmats.2019.00123 is OK
- 10.1016/j.cpc.2009.06.022 is OK
- 10.1103/physrev.46.618 is OK
- 10.1063/1.3043729 is OK
- 10.1016/s0009-2614(89)87395-6 is OK
- 10.1063/1.455269 is OK
- 10.1063/1.4977994 is OK
- 10.1088/1367-2630/14/5/053020 is OK
- 10.1021/acs.jpclett.0c00481 is OK
- 10.1021/acs.jpclett.8b03679 is OK
- 10.1063/1.5055706 is OK
- 10.1103/physrevb.101.165138 is OK
- 10.1021/acs.jctc.7b00049 is OK
- 10.1021/acs.jctc.5b00422 is OK
- 10.1103/physrevlett.117.133002 is OK
- 10.1021/acs.jpca.9b03976 is OK
- 10.1063/1.1388045 is OK
- 10.1063/1.4985878 is OK
- 10.1103/physrevb.101.241113 is OK
- 10.1103/physrevlett.52.1830 is OK
- 10.1103/physrevb.54.1703 is OK
- 10.1103/physrevb.73.035404 is OK
- 10.1016/0022-3093(95)00355-x is OK
- 10.1039/9781849734790-00168 is OK
- 10.1007/128_2014_600 is OK
- 10.1021/acs.jctc.9b00957 is OK
- 10.1021/acs.jctc.9b00511 is OK
- 10.1021/acs.jctc.0c01129 is OK
- 10.1063/1.464746 is OK
- 10.1103/physrevb.59.1758 is OK
- 10.1088/1367-2630/17/9/093020 is OK
- 10.1002/jcc.23284 is OK
- 10.1063/1.1679012 is OK
- 10.1016/0009-2614(93)87156-w is OK
- 10.1021/acs.jctc.0c00101 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by Evmoerman** [comment]
@awvwgk I have fixed the compilation issues with GCC-11. Tell me please, if you still experience issues with this GCC version.

**Comment by awvwgk** [comment]
Thanks, I'll try to build everything ASAP, latest next weekend ;).

**Comment by wcwitt** [comment]
@Evmoerman I don't have FHI-aims or CC4S either, sorry. Would you mind getting me access? I'm at Cambridge in the [Materials Theory Group](https://www.mtg.msm.cam.ac.uk/People/dr-w-chuck-witt).

**Comment by Evmoerman** [comment]
@wcwitt Of course. I will send some email in a minute.

**Comment by awvwgk** [comment]
Updated my review in the description above. In summary, the documentation is a bit brief

- guidelines for providing feedback and contributing patches would be good:
  https://gitlab.com/moerman1/fhi-cc4s/-/issues/3
- installation instruction for non-Linux platforms:
  https://gitlab.com/moerman1/fhi-cc4s/-/issues/4
- the user documentation could use tutorial and/or recipe:
  https://gitlab.com/moerman1/fhi-cc4s/-/issues/5

**Comment by Evmoerman** [comment]
@awvwgk Thank you very much for your comments. I have already been thinking about your 1st and the 3rd point. However, - as you also state in the issue - I'm not sure if it is possible/makes sense to offer installation on windows because, as far as I know, FHI-aims and CC4S both assume a Linux environment. However, I will have a look and try to add documentation/tutorials where possible.

Thanks again.

**Comment by awvwgk** [comment]
In any case, clearly stating that the project is only officially supported on Linux (or Unix) systems will help to set the right expectations for users.

**Comment by Evmoerman** [comment]
Got it. I will certainly do that.

**Comment by rkurchin** [comment]
👋 @wcwitt, just checking in whether you've been able to start your review yet!

**Comment by wcwitt** [comment]
I haven't, apologies, but I will ASAP.

**Comment by whedon** [comment]
:wave: @awvwgk, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @wcwitt, please update us on how your review is going (this is an automated reminder).

**Comment by wcwitt** [comment]
I've managed to install the interface package, and I'm working on installing the rest (it took a while to get FHI-aims - thanks @Evmoerman for your help with that!).

My main initial reaction is that a full example is really necessary (also mentioned by @awvwgk https://gitlab.com/moerman1/fhi-cc4s/-/issues/5).

The paper and documentation also suggest it should be straightforward to use this interface with codes other than FHI-aims, but I would prefer to see evidence if possible. I can see that real effort went into this code, which I'd like to see rewarded, but if it's mainly an extension of FHI-aims I'm not certain it meets the eligibility criteria. Here is an issue to discuss (https://gitlab.com/moerman1/fhi-cc4s/-/issues/6).

**Comment by Evmoerman** [comment]
@wcwitt Currently, CC-aims is only used in combination with FHI-aims. This is merely the result of me working at the Fritz-Haber-Institute, in the group which develops the FHI-aims code. 
However, there is nothing in CC-aims, which assumes that FHI-aims is the ab-initio code being used. The quantities required by CC-aims can be provided by any code which meets the criteria stated in the documentation.
The idea is, that once CC-aims is published other localized-orbital codes will either interface to it and/or use it as a template for similar interfaces.

**Comment by awvwgk** [comment]
>  However, there is nothing in CC-aims, which assumes that FHI-aims is the ab-initio code being used. The quantities required by CC-aims can be provided by any code which meets the criteria stated in the documentation.
>  The idea is, that once CC-aims is published other localized-orbital codes will either interface to it and/or use it as a template for similar interfaces.

On that note it might be worth to compare to Crystal/Cryscor ([_Phys. Chem. Chem. Phys._, **2012**,14, 7615-7628](https://doi.org/10.1039/C2CP23927B)) since the general design of FHI-aims/CC-aims seems similar on the first glance. Also, what (open source) QC programs would be primary targets for interfacing with CC-aims?

**Comment by Evmoerman** [comment]
Indeed, CRYSTAL would be a great candidate, as it provides a localized basis+density fitting but - as far as I remember - only offers MP2. Hence, an interface to CC4S would expand its range of functionality immensely. 
Another code, which would very likely benefit from an interface like CC-aims is QuantumATK, which is not open-source but free-ware for academic use. 
It is hard to come up with a list of truly open-source quantum-chemistry programs, as they are comparably rare. One such code, is PySCF. While PySCF offers a range of beyond-DFT methods, I personally experienced them to be unbelievably memory-demanding for bigger calculations (even for HPC standards), while the same calculations could be reasonably performed with CC4S. 
In this case, other codes could benefit from CC-aims/CC4S even though the codes themselves include beyond-DFT methods as I point out in the final paragraph of my manuscript.

**Comment by rkurchin** [comment]
Hi all, just wanted to check in on the status here. Seems there are at least those two issues open in the repo: one regarding a full example with FHI-aims, and one regarding interfacing to another code as this is a "selling point" of the package. From my perspective, the first one should definitely be fulfilled before acceptance. The second one would definitely be nice as well, but I'd like perspective both from reviewers on how necessary they see it to be and from @Evmoerman on how much work this would represent...

**Comment by Evmoerman** [comment]
@rkurchin Yes, I fully agree. While I cannot provide an example except for FHI-aims, I of course will include a tutorial/example page on the website and I also believe that contributing guidelines, like @awvwgk suggested, are also necessary.
I'm sorry I haven't taken care of these issues yet, because I'm currently visiting the CC4S Group in Vienna, where I help developing. I am certainly aware of the to do list, but the last 2 weeks were a bit stressful because I want to make the most of my short visit here.
I will take care of the review issues as soon as I can.
Sorry again for the delay.

**Comment by wcwitt** [comment]
Thanks - my main update is I've managed to install FHIaims with CC-aims and am looking forward to the example. @Evmoerman, I doubt any of us are worried about delays so all good. @rkurchin, I will think about your question.

**Comment by awvwgk** [comment]
@Evmoerman, take your time and make the most out of your visit.

I already tried setting up a calculation myself, but since I'm neither familiar with FHI-aims nor CC4S it takes some time to find the right information in the manual. Some points I stumbled upon:

- I had to learn yet again a new geometry input format (this is actually on FHI-aims)
- FHI-aims manual is quite detailed on a lot of things, however it doesn't mention CC4S or CC-aims? (version 210716_2)
- to get any insights on running FHI-aims/CC-aims I had to dive into the source and search for the keywords advertised in the [docs](https://moerman1.gitlab.io/fhi-cc4s/for_FHIaims_users/#running-cc-aims-in-fhi-aims) to see whether they actually exist
- the only other alternative is to search through the [test-inputs](https://gitlab.com/moerman1/fhi-cc4s/-/blob/master/tests/test-inputs/H2O-simple/control.in) in the repo and just run one of those

I didn't really try to create an input myself yet and run a somewhat bigger test calculation, mainly because this would involve a lot of trial and error based on few information I could gather from the test-suite/source code.

**Comment by Evmoerman** [comment]
@awvwgk Yea, I can see how this may be difficult at first. While invoking a CC-aims calculation/post-processing is controlled by only 1 or 2 keywords in FHI-aims, it isn't as trivial setting up the rest of the input (including the geometry.in you mentioned). 
While the test-inputs contain a variety of suitable inputs for FHI-aims+CC-aims, and I will certainly use one/some of these for the tutorial section, you are totally right that a separate chapter in the CC-aims documentation should exist for that.
Finally, the CC-aims-related keywords are not yet documented in the FHI-aims manual because I was suspecting that the details about CC-aims may change in the course of the JOSS review, so that I didn't want to expose new potential users of FHI-aims to a functionality, which is not yet in its final stage and tried to document it as precisely as possible in the CC-aims documentation.

**Comment by Evmoerman** [comment]
Hey @rkurchin. My visit here is until end of this month, but I actually started already (slowly) to fix issues with CC-aims. 
This week (fingers crossed) the CC4S repository goes public and I am currently adapting the file formats and the CI/CD pipeline to the new CC4S infrastructure. 
After the testing suite is stable and compatible again with the released CC4S, I hopefully can take care of the documentation/tutorial issues.

**Comment by Evmoerman** [comment]
Hey guys, @rkurchin , @awvwgk , @wcwitt , I have now adapted the interface to work with the **now public** version of CC4S, which can be found [here](https://gitlab.cc4s.org/cc4s/cc4s).
The tests and pipelines work and use the newest version of CC4S.
If nothing breaks in the meantime, I hope that I can start to take care of the issues opened so far.

**Comment by wcwitt** [comment]
Sounds great. Once there is an example, I'll work through it and finish with any other comments.

**Comment by awvwgk** [comment]
Just tried to compile CC4S again and failed with a more recent version of GCC. The overall experience of building CC4S is a bit unsettling, having a makefile invoke CMake and autotools to build dependencies seems rather fragile to me, also I have a version of yaml-cpp around which I prefer to use rather than compile yet another one. This is not really on CC-aims but more on CC4S.

I really appreciate a high-level build system (meson, CMake, or even autotools) over a bunch of makefiles. Chances that a makefile just fails are rather high (at least my success rate with makefiles has been rather low so far).

**Comment by Evmoerman** [comment]
@awvwgk Thanks for your assessment. I, of course, cannot guarantee that the current installation infrastructure will change in the near future but I will forward your comment to the CC4S development team. If possible, it would probably also be a good idea to open an issue on the public cc4s repo.
Is it okay if I include your e-mail address, so that someone can contact you for that issue?

**Comment by awvwgk** [comment]
I can't open issues at the CC4S GitLab, so feel free to go ahead and open an issue on my behalf.

Another thing I would appreciate would be a released version of CC4S (maybe also CTF). A C++/MPI program should be easy enough to package on conda-forge, which in turn would simplify the CC-aims installation because dependencies could be installed from a package manager (for HPC usage it is of course preferable to compile from source and tune for the architecture).

**Comment by Evmoerman** [comment]
I see. I assume that Issues have not been enabled yet for everyone so early after the code release. I have forwarded your bug report+installation recommendations to the devs. I will also inform them about your advice to package CC4S.

**Comment by Evmoerman** [comment]
@awvwgk, I just got word back:
1. Following your bug report of the issue opened on the CC-aims project, the compilation error lies in the CTF library, not in CC4S. This will be investigated probably this week.
2. Automation of the installation/building environment: A cmake-solution is also desired by the developers, and this may be adapted in the following versions of CC4S. For now, however, this is not on the top of the priority list. I hope that's not a too pressing issue.
3. Utilization of a custom yaml-cpp: I'm afraid it is not entirely clear, what the issue there is. Is CC4S not compiling with your version of the yaml-library? Or are you missing a straight-forward possibility to make CC4S even use your own yaml-library? You would need to elaborate on that.
4. Issues on the public CC4S-project for non-members will be enabled the following days (perhaps today).

**Comment by awvwgk** [comment]
Thanks a lot.

> Utilization of a custom yaml-cpp: I'm afraid it is not entirely clear, what the issue there is. Is CC4S not compiling with your version of the yaml-library? Or are you missing a straight-forward possibility to make CC4S even use your own yaml-library? You would need to elaborate on that.

I was surprised to get a full build of `yaml-cpp` when starting the makefile, while I have an installation of `yaml-cpp` already on my system. I could probably hack this myself into the makefile (not a big fan of this solution), but I would expect a build system to prefer installed versions over vendored ones by default.

**Comment by Evmoerman** [comment]
I'm afraid for now there is only the "hacky" way to force CC4S to use custom dependencies.
Furthermore, if I was informed correctly, you can now register on the public CC4S repo, which allows you to open issues.

**Comment by rkurchin** [comment]
Hi all, just wanted to check in on this – reviewers, have you been able to get the software running to check the functionality claims?

**Comment by wcwitt** [comment]
I believe we're still waiting on example input (https://gitlab.com/moerman1/fhi-cc4s/-/issues/5), although happy to be corrected

**Comment by Evmoerman** [comment]
Yes, I am working on the tutorial at the moment. Returned from Vienna this week.

**Comment by Evmoerman** [comment]
@awvwgk @wcwitt @rkurchin I have finally added a tutorial, which can be found on the "For FHI-aims users" page of the documentation.

I also addressed all other current issues as far as I see, including contributing guidelines and specification of the required OS.
If I forgot something, please tell me so.

**Comment by wcwitt** [comment]
Thanks, I'll have a look this week.

**Comment by rkurchin** [comment]
Hi @wcwitt and @awvwgk, have you had a chance to take a look at the changes @Evmoerman made?

**Comment by wcwitt** [comment]
Thanks for the reminder. I've looked briefly but need some more time as well.

**Comment by Evmoerman** [comment]
Hey @rkurchin, I'm currently trying to submit the manuscript of this project at arxiv. However, it requires me to submit it a .tex file. Due do the specifics of JOSS, however, I'm not able to do that. 
This is probably a quite frequently occuring question as you also actively encourage people to publish the manuscript on arvix.

Alternatively, it would also be okay, if JOSS provides some kind of identifiers (like DOIs) for manuscripts under review.
My supervisor is just asking me to provide a reference to my work, which refers interested readers to results of his group.

**Comment by wcwitt** [comment]
Just to say I'll aim to finish my next round of review this week.

And I expect @rkurchin will be able to give a more definitive answer to your question @Evmoerman, but I know there is some relevant discussion here: https://github.com/openjournals/joss/issues/132. Since you'd be uploading a preprint to arxiv, I don't think you'd want the JOSS logo or styling, so you could also try converting the markdown to .tex with `pandoc`.

**Comment by Evmoerman** [comment]
@rkurchin @wcwitt Thank you for that recommendation. I saw that discussion, however the solutions presented there seemed rather complicated. Hence, I hoped there would be an easier way. 
I tried converting the md-file via pandoc using the template (here)[https://github.com/openjournals/whedon/blob/master/resources/latex.template]. For some reason, however, the references/bibliography is not correctly resolved, as can be seen in the attached file.

**Comment by rkurchin** [comment]
I actually haven't dealt with this myself, so I'm not 💯 certain what the easiest solution would be. However, the one posed in [this comment](https://github.com/openjournals/joss/issues/132#issuecomment-890440692) on that thread looks reasonably clear?

I agree that this is a sensible thing to want to do, so do let me know if you have further trouble...

**Comment by Evmoerman** [comment]
Thanks for your response. I'm not familiar with GitHub actions, except for the fact that is has been superseded by GitLab CI. Can I do the same thing described in that comment on GitLab, where my code/manuscript/project lies, or do I have to create a small GitHub project for that?

**Comment by rkurchin** [comment]
Ah, I see...well I did find [this post](https://tomasvotruba.com/blog/how-can-we-use-github-actions-in-gitlab/) on using GitHub Actions in GitLab CI, but given you really only need this one action to run once (or maybe a couple times), it might be more straightforward to just make a GitHub repo and add it as a second remote for your local version...

Sorry that this is a bit trickier than anticipated!

**Comment by Evmoerman** [comment]
Yea, so many workaround for a seemingly ordinary task. I appreciate your recommendation and your comment on the issue. I will try to resolve it on GitHub then. Thank you very much.

**Comment by Evmoerman** [comment]
The github approach worked like a charm. Just a minor question wrt the logo. Is it better if I leave out the logo as the work is under review, or is that irrelevant in this case?

**Comment by rkurchin** [comment]
I would say leave it out. If that looks strange or something, maybe just use a different style file?

—————————————
Rachel C. Kurchin, PhD
Postdoctoral Researcher (formerly MolSSI and MFI Fellow)
Carnegie Mellon Mechanical Engineering and Materials Science
https://rkurchin.github.io

**Comment by Evmoerman** [comment]
No, it wouldn't look strange (in my opinion). I just didn't want to wrongly advertise the work as JOSS-approved. Just leaving out the logo will look just fine.

**Comment by rkurchin** [comment]
Sounds good. You are of course free (and encouraged!) to include the logo once the work is accepted! :)

—————————————
Rachel C. Kurchin, PhD
Postdoctoral Researcher (formerly MolSSI and MFI Fellow)
Carnegie Mellon Mechanical Engineering and Materials Science
https://rkurchin.github.io

**Comment by wcwitt** [comment]
I've finished my review and am pleased to recommend publication, provided @awvwgk is satisfied and the following are addressed

* Example output is uploaded: https://gitlab.com/moerman1/fhi-cc4s/-/issues/5
* Claims about using other codes: https://gitlab.com/moerman1/fhi-cc4s/-/issues/6
* Comments on the paper: https://gitlab.com/moerman1/fhi-cc4s/-/issues/8

**Comment by wcwitt** [comment]
(Also, @awvwgk, I think I accidentally checked and/or unchecked some boxes in your checklist - apologies!)

**Comment by Evmoerman** [comment]
@wcwitt Thank you for your assessment. I'm glad that eventually the step-by-step tutorial worked for you. I will see if I can rephrase the statements concerning interfacing other codes to be more precise. While I certainly believe that what is said in the manuscript is correct both in principal and practically, I agree, that a more cautious wording may be necessary.
The other things you mentioned are more straight-forward and I will probably have fixed them by the end of the week.

**Comment by rkurchin** [comment]
Hi @Evmoerman, just wanted to check in on the status of the changes you mentioned in the prior comment!

**Comment by Evmoerman** [comment]
Hey @rkurchin. Except for the point addressing the interfaceability with other codes, I fixed all the issues the same day I believe.
That last change I have put aside for the moment until I have the comments by Sebastian. I figured that this change is not so urgent as it only involves the manuscript and not the code.

**Comment by rkurchin** [comment]
Got it, understood. @awvwgk, any updates from your end?

**Comment by awvwgk** [comment]
Sorry for the delay, I'm currently occupied writing my thesis and wasn't able to devote time for this review, unfortunately.

The software paper is well written and seems sufficiently supported by the software. I have no complaints about this part.

Regarding the software documentation, I'm surprised to find screenshots rather than code block in the documentation, which makes the example more difficult to follow than necessary (and screenshots are not accessible). Also, rather than copying file from the repository, which might change, including them verbatim in the documentation would be preferable.

I'll try to allocate some time this week to compile the software stack again and run through the examples.

**Comment by Evmoerman** [comment]
Haha yea, I didn't notice that at all. It will be no problem to redo those. We already had the discussion about code blocks vs screenshots(https://gitlab.com/moerman1/fhi-cc4s/-/issues/5). Essentially, I used screenshots here because I wanted to emphasize in which directory the user should be each step of the time and how the copied input files look in side. The commands used in this screenshots are mostly if not exclusively elementary bash commands like mkdir, ls or cd. 
In other words, there aren't any things the user needs to enter beyond basic bash commands. Otherwise I would have used code blocks more.

I thought about including the FHI-aims input as code blocks. In the case of geometry.in this would be no problem, control.in however is a comparably long file (due to the basis set information). I tried out including control.in as a code block, however the page looked very cluttered by the noticeably larger code block. Hence, I dropped that idea for aesthetic reasons.

**Comment by rkurchin** [comment]
Hi again @awvwgk, just checking in again!

**Comment by rkurchin** [comment]
🔔 ping again @awvwgk!

**Comment by awvwgk** [comment]
Thanks for the reminder, last months have been busy for me with study related work and interviewing.

---

I had a look on the CC-aims tutorial and unfortunately failed based on the given instructions. There is an issue that no stable FHI-aims release currently supports the shown example and a development version has to used from the latest git source. This seems like a moving target and an issue potential users will for sure run into.

I know that release schedules in research software can be quite irregular, however it would be good if this behavior is at least documented. The requirements in CC-aims should clearly state which version of FHI-aims are known to work.

Another issue is that the FHI-aims manual, for the last release I checked, was silent about the flags for using CC-aims, this would be good to have documented in detail in the CC-aims or (better and) the FHI-aims docs.

**Comment by Evmoerman** [comment]
I agree with your assessment: I expected the next release to be sooner, so that I didn't think your first finding would originally  be a problem. However, until then I will add a comment about it in the CC-aims documentation.
Wrt. your second point, I originally didn't add the new functionalities to the FHI-aims documentation on purpose, because I considered it unfinished until it's published/officially released. However, judging from the close-to-finished review (assuming that you don't run into unexpected complications while testing), there is no reason not to add the CC-aims-related keywords to the FHI-aims manual. I will do so, hopefully, this week.

**Comment by awvwgk** [comment]
I see I already mentioned this a while ago, see https://github.com/openjournals/joss-reviews/issues/4040#issuecomment-1028435682.

> * I had to learn yet again a new geometry input format (this is actually on FHI-aims)
> 
> * FHI-aims manual is quite detailed on a lot of things, however it doesn't mention CC4S or CC-aims? (version 210716_2)
> 
> * to get any insights on running FHI-aims/CC-aims I had to dive into the source and search for the keywords advertised in the [docs](https://moerman1.gitlab.io/fhi-cc4s/for_FHIaims_users/#running-cc-aims-in-fhi-aims) to see whether they actually exist
> 
> * the only other alternative is to search through the [test-inputs](https://gitlab.com/moerman1/fhi-cc4s/-/blob/master/tests/test-inputs/H2O-simple/control.in) in the repo and just run one of those

I still think this is a blocker which must be addressed in either CC-aims and FHI-aims. We don't have to wait for an FHI-aims release here with the respective documentation update and bug fixes, if the CC-aims repository provides a sufficient detailed documentation on the available keywords in FHI-aims and the version constraints.

**Comment by Evmoerman** [comment]
I'm not sure if you have taken this into account, but there is a detailed documentation of the CC-aims related keywords for FHI-aims on the CC-aims documentation under  "For FHI-aims users/Running CC-aims in FHI-aims".

**Comment by awvwgk** [comment]
The crucial information about the status in FHI-aims and the required version is missing for example.

**Comment by Evmoerman** [comment]
Yes, that is correct. I'm going to add that. I was referring to the second half of your objections.

**Comment by Evmoerman** [comment]
I added a disclaimer about using the development version of FHI-aims to the "For FHI-aims users"-section of the documentation.

**Comment by awvwgk** [comment]
Thanks, I think this should address all my remaining concerns.

**Comment by rkurchin** [comment]
@editorialbot generate pdf

**Comment by rkurchin** [comment]
Thanks everyone! Authors, I'll do an editorial pass over the manuscript and send any comments shortly. In the meantime, the next steps for you are:

1. Merge any and all changes from this review into your main branch and **issue a new version tag**. _(If you want to merge in the paper, you may, but it is not required that the actual manuscript live into the repo in perpetuity since JOSS will host it and you can simply add a badge link or whatever you like. But the actual changes to software and docs do need to be merged!)_
2. Create a DOI for the contents of the repo at the same commit corresponding to that version tag, e.g. using figshare or Zenodo. Please make sure that the metadata (version number, title, author list, etc.) match those of your manuscript.
3. Post a comment here with the version number and DOI.

**Comment by rkurchin** [comment]
Some fairly minor editorial comments:
* line 24-25: "allows to calculate" is ungrammatical, rephrase to either "...allows one to calculate..." or "...enables calculation of..."
* 25: I think "ground-state" should be hyphenated
* 36: boron nitride does not need to be hyphenated
* 45: remove comma after "groups"
* 46: perform -> performs
* 55-56: "straightforward" is one word
* 64: change "which" to "that" (or add comma before "which")
* 62-69: I would remove the "on the one hand" and "on the other hand" as that construction is normally used when comparing advantages/disadvantages, whereas you are just pointing out distinct features
* 76: I would remove the comma and change "which" to "that" in this case
* 78: remove comma after "memory-saving"
* 84 and 86: no need to capitalize "quantum"
* 84-85: Rephrase to "...localized basis, an RI scheme is also needed (and thus may need to be implemented)."
* 88: remove comma after "areas"

**Comment by Evmoerman** [comment]
@whedon generate pdf

**Comment by Evmoerman** [comment]
@editorialbot generate pdf

**Comment by rkurchin** [comment]
I'm not very familiar with either GitLab or figshare, to be honest. Have you tried Zenodo? If that doesn't work, either, I can tag in the EiC team for some help.

**Comment by Evmoerman** [comment]
I could also just manually zip a local copy of the current gitlab repo and upload that to figshare (i.e without any synchronization). Would that be alright?

**Comment by Evmoerman** [comment]
DOI has been generated by zenodo: version is 1.0.1 and doi is 10.5281/zenodo.6658117

**Comment by rkurchin** [comment]
@editorialbot set 1.0.1 as version

**Comment by rkurchin** [comment]
@editorialbot set 10.5281/zenodo.6658117 as archive

**Comment by rkurchin** [comment]
@editorialbot recommend-accept

**Comment by Evmoerman** [comment]
Just to make sure I understand: It's not me, who is supposed to accept the final paper right?

**Comment by danielskatz** [comment]
No, it's the @openjournals/joss-eics on duty this week who will do this

**Comment by Kevin-Mattheus-Moerman** [comment]
- I checked the archive meta-data and all looks good. :+1: 
- I checked the version tag for the review and archive and verified they match  :+1:
- I proofread the paper and it looks fine :+1: 

I will now proceed to process this work for acceptance in JOSS

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot accept

**Comment by Kevin-Mattheus-Moerman** [comment]
@Evmoerman congratulations on getting this work published in JOSS! 

@rkurchin thanks for editing and thank you @awvwgk and @wcwitt for your stellar review efforts! :tada:
